# Ejercicios de repaso FP

## CÓMO RESOLVER LOS EJERCICIOS
A la hora de resolver los ejercicios, es muy importante que, sobre todo al principio, antes de poneros a programar dediquéis unos minutos a pensar en lo que os pide el ejercicio y cómo estructurar el código. Para ello os propongo el siguiente proceso: 

* Definid, en dos líneas, qué os está pidiendo el ejercicio
* Identificad las ENTRADAS del programa y de dónde se obtienen
* Identificad qué tiene que PROCESAR el programa
* Identificad qué SALIDA tiene que producir el programa

Sólo una vez que tenéis claro todo esto os debéis poner a programar. 

Happy programmaing! 

## Recordatorio: uso de la ayuda del Python

Recordad que podéis en cualquier momento invocar el método help() de Python para obtener ayuda sobre métodos, objetos, librerías, etc. 

Muestra la ayuda de los objetos lista, diccionario, conjunto y tupla, y contesta a las siguientes preguntas:
* ¿Qué métodos oferta la clase lista?
* ¿Qué clases ofertan el método count? 
* ¿Qué devuelve el método index() de tuple si no encuentra el valor que se busca?
* El método list.insert(), ¿inserta el objeto antes o después de la posición que indica el índice que se pasa como parámetro? ¿Qué ocurre si se le pasa un índice fuera de rango?
Veamos algunos ejemplos de invocación de la ayuda: 
<pre>
help(list)
help(dict)
help(set)
help(tuple)
help(list.count)
</pre>

In [7]:
help(list.count)

Help on method_descriptor:

count(self, value, /)
    Return number of occurrences of value.



# 5a.- LISTAS

## Ejercicio 5a.0: Phras-o-matic (Libro Head First Learn 2 Code)
Escribe un programa que, dadas tres listas con verbos, adjetivos y nombres llenos de 'palabras de moda' (buzz words), cree diez eslóganes potenciales para tu próximo proyecto. 

Ejemplo de salida: 
* Sync B-to-B Splash Page
* Sync Siloed Early Adopter
* 30,000 foot Oriented Early Adopter
* Gamify Oriented Splash Page
* Target B-to-B Paradigm
* 24/7 Oriented Paradigm
* 30,000 foot B-to-B Low-hanging Fruit
* Target Oriented Low-hanging Fruit
* Target A/B Tested Paradigm
* Sync A/B Tested Early Adopter

In [3]:


verbs = ['Leverage', 'Sync', 'Target',
         'Gamify', 'Offline', 'Crowd-sourced',
         '24/7', 'Lean-in', '30,000 foot']

adjectives = ['A/B Tested', 'Freemium',
              'Hyperlocal', 'Siloed', 'B-to-B',
              'Oriented', 'Cloud-based', 
              'API-based']


nouns = ['Early Adopter', 'Low-hanging Fruit',
         'Pipeline', 'Splash Page', 'Productivity',
         'Process', 'Tipping Point', 'Paradigm']



## Ejercicio 5a.1

Existe un formato de fichero utilizado para mantener colecciones de mensajes de correo electrónico llamado mbox (https://en.wikipedia.org/wiki/Mbox)
Dentro de esos ficheros, cada mensaje incluye una serie de líneas generadas por un filtro software estadístico de spam de código abierto llamado DSPAM, https://help.ubuntu.com/community/Postfix/Dspam)


En ese formato, la primera línea indica si ese correo es spam (X-DSPAM-Result: Innocent), la segunda indica cuándo fue analizado (X-DSPAM-Processed: Tue Jun 16 20:25:03 2009),  la tercera indica el grado de confianza (entre 0 y 1) en el veredicto (inocente/spam) (X-DSPAM-Confidence: 1.0000) y la cuarta línea indica la probabilidad de que el correo sea spam (X-DSPAM-Probability: 0.0023)


A partir de un fichero de mail con formato mbox que contiene numerosos mensajes, se ha extraído la lista de cadenas almacenada en la variable lista_conf. 

Esta lista guarda el nivel de confianza de los distintos mensajes. A partir de ella, se pide calcular el nivel medio de confianza en que los mensajes de ese fichero estén bien reportados como legítimos o SPAM. Redondead el resultado a cuatro decimales.  

Nota: la salida os debería dar 0.7507


In [43]:
lista_conf=['X-DSPAM-Confidence:0.8475', 'X-DSPAM-Confidence: 0.6178', 
            'X-DSPAM-Confidence:   0.6961', 'X-DSPAM-Confidence: 0.7565', 
            'X-DSPAM-Confidence: 0.7626', 'X-DSPAM-Confidence: 0.7556', 
            'X-DSPAM-Confidence: 0.7002', 'X-DSPAM-Confidence: 0.7615', 
            'X-DSPAM-Confidence: 0.7601', 'X-DSPAM-Confidence: 0.7605', 
            'X-DSPAM-Confidence: 0.6959', 'X-DSPAM-Confidence: 0.7606', 
            'X-DSPAM-Confidence: 0.7559', 'X-DSPAM-Confidence: 0.7605', 
            'X-DSPAM-Confidence: 0.6932', 'X-DSPAM-Confidence: 0.7558', 
            'X-DSPAM-Confidence: 0.6526', 'X-DSPAM-Confidence: 0.6948', 
            'X-DSPAM-Confidence: 0.6528', 'X-DSPAM-Confidence: 0.7002', 
            'X-DSPAM-Confidence: 0.7554', 'X-DSPAM-Confidence: 0.6956', 
            'X-DSPAM-Confidence: 0.6959', 'X-DSPAM-Confidence: 0.7556', 
            'X-DSPAM-Confidence: 0.9846', 'X-DSPAM-Confidence: 0.8509', 
            'X-DSPAM-Confidence: 0.9907']


def media_confianza(lista):
    lista1=[]
    for cadena in lista:
        #num=cadena[20:26]
        #lista_elem=cadena.split(":")
        #num1=float(lista_elem[1])
        basura,num=cadena.split(":")
        num1=float(num)
        lista1.append(num1)
    print(lista1)
    valor=sum(lista1)/len(lista1)
    return(round(valor,4))

media_confianza(lista_conf)

[0.8475, 0.6178, 0.6961, 0.7565, 0.7626, 0.7556, 0.7002, 0.7615, 0.7601, 0.7605, 0.6959, 0.7606, 0.7559, 0.7605, 0.6932, 0.7558, 0.6526, 0.6948, 0.6528, 0.7002, 0.7554, 0.6956, 0.6959, 0.7556, 0.9846, 0.8509, 0.9907]


0.7507

In [41]:
lista_conf=['X-DSPAM-Confidence: 0.8475', 'X-DSPAM-Confidence: 0.6178', 
            'X-DSPAM-Confidence: 0.6961', 'X-DSPAM-Confidence: 0.7565', 
            'X-DSPAM-Confidence: 0.7626', 'X-DSPAM-Confidence: 0.7556', 
            'X-DSPAM-Confidence: 0.7002', 'X-DSPAM-Confidence: 0.7615', 
            'X-DSPAM-Confidence: 0.7601', 'X-DSPAM-Confidence: 0.7605', 
            'X-DSPAM-Confidence: 0.6959', 'X-DSPAM-Confidence: 0.7606', 
            'X-DSPAM-Confidence: 0.7559', 'X-DSPAM-Confidence: 0.7605', 
            'X-DSPAM-Confidence: 0.6932', 'X-DSPAM-Confidence: 0.7558', 
            'X-DSPAM-Confidence: 0.6526', 'X-DSPAM-Confidence: 0.6948', 
            'X-DSPAM-Confidence: 0.6528', 'X-DSPAM-Confidence: 0.7002', 
            'X-DSPAM-Confidence: 0.7554', 'X-DSPAM-Confidence: 0.6956', 
            'X-DSPAM-Confidence: 0.6959', 'X-DSPAM-Confidence: 0.7556', 
            'X-DSPAM-Confidence: 0.9846', 'X-DSPAM-Confidence: 0.8509', 
            'X-DSPAM-Confidence: 0.9907']


for e in lista_conf:
    print(e[20:26])

0.8475
0.6178
0.6961
0.7565
0.7626
0.7556
0.7002
0.7615
0.7601
0.7605
0.6959
0.7606
0.7559
0.7605
0.6932
0.7558
0.6526
0.6948
0.6528
0.7002
0.7554
0.6956
0.6959
0.7556
0.9846
0.8509
0.9907


## Ejercicio 5a.2
Dada la lista de remitentes contenida en la variable lista_from, se pide imprimir la lista de correos únicos desde donde se han enviado los mensajes

Pista: una estrategia sencilla para quedarte con correos únicos (aplicando sólo lo que hemos visto hasta el momento) es que, una vez extraídos los correos en una lista, añadas a otra lista los correos que aún no estén en ella. Otra posibilidad es ordenar la lista y quedarse con la primera ocurrencia de cada uno. Tb puedes convertir la lista a un conjunto para quitar duplicados. Por último, puedes convertir la lista en claves de un diccionario (que son únicas por definición). Más info en http://blog.elcodiguero.com/python/eliminar-objetos-repetidos-de-una-lista.html


![Ejemplos de salida](ejemplosalida2.png)


In [8]:
lista_from=['From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008', 
            'From: stephen.marquard@uct.ac.za', 
            'From louis@media.berkeley.edu Fri Jan  4 18:10:48 2008', 
            'From: louis@media.berkeley.edu', 
            'From zqian@umich.edu Fri Jan  4 16:10:39 2008', 
            'From: zqian@umich.edu', 
            'From rjlowe@iupui.edu Fri Jan  4 15:46:24 2008', 
            'From: rjlowe@iupui.edu', 
            'From zqian@umich.edu Fri Jan  4 15:03:18 2008', 
            'From: zqian@umich.edu', 
            'From rjlowe@iupui.edu Fri Jan  4 14:50:18 2008', 
            'From: rjlowe@iupui.edu', 'From cwen@iupui.edu Fri Jan  4 11:37:30 2008', 
            'From: cwen@iupui.edu', 'From cwen@iupui.edu Fri Jan  4 11:35:08 2008', 
            'From: cwen@iupui.edu', 'From gsilver@umich.edu Fri Jan  4 11:12:37 2008', 
            'From: gsilver@umich.edu', 'From gsilver@umich.edu Fri Jan  4 11:11:52 2008', 
            'From: gsilver@umich.edu', 'From zqian@umich.edu Fri Jan  4 11:11:03 2008', 
            'From: zqian@umich.edu', 'From gsilver@umich.edu Fri Jan  4 11:10:22 2008', 
            'From: gsilver@umich.edu', 'From wagnermr@iupui.edu Fri Jan  4 10:38:42 2008', 
            'From: wagnermr@iupui.edu', 'From zqian@umich.edu Fri Jan  4 10:17:43 2008', 
            'From: zqian@umich.edu', 'From antranig@caret.cam.ac.uk Fri Jan  4 10:04:14 2008', 
            'From: antranig@caret.cam.ac.uk', 
            'From gopal.ramasammycook@gmail.com Fri Jan  4 09:05:31 2008', 
            'From: gopal.ramasammycook@gmail.com', 
            'From david.horwitz@uct.ac.za Fri Jan  4 07:02:32 2008', 
            'From: david.horwitz@uct.ac.za', 
            'From david.horwitz@uct.ac.za Fri Jan  4 06:08:27 2008', 
            'From: david.horwitz@uct.ac.za', 'From david.horwitz@uct.ac.za Fri Jan  4 04:49:08 2008', 
            'From: david.horwitz@uct.ac.za', 'From david.horwitz@uct.ac.za Fri Jan  4 04:33:44 2008', 
            'From: david.horwitz@uct.ac.za', 
            'From stephen.marquard@uct.ac.za Fri Jan  4 04:07:34 2008', 
            'From: stephen.marquard@uct.ac.za', 
            'From louis@media.berkeley.edu Thu Jan  3 19:51:21 2008', 
            'From: louis@media.berkeley.edu', 
            'From louis@media.berkeley.edu Thu Jan  3 17:18:23 2008', 
            'From: louis@media.berkeley.edu', 
            'From ray@media.berkeley.edu Thu Jan  3 17:07:00 2008', 
            'From: ray@media.berkeley.edu', 'From cwen@iupui.edu Thu Jan  3 16:34:40 2008', 
            'From: cwen@iupui.edu', 'From cwen@iupui.edu Thu Jan  3 16:29:07 2008', 
            'From: cwen@iupui.edu', 
            'From cwen@iupui.edu Thu Jan  3 16:23:48 2008', 'From: cwen@iupui.edu']

#aquí va tu código


## Ejercicio 5a.3
Implementa  una función que devuelva el número de palabras de la cadena contenida en la variable texto. 

Nota: Fíjate en que el texto tiene espacios en blanco extras, \n, que tenéis que tener en cuenta a la hora de contar las palabras. Con el comando repr(texto) podéis ver todos esos caracteres ocultos. El resultado debería ser 143

Una vez calculado el número de palabras, crea otra función que te devuelva la longitud media de las palabras contenidas en el texto redondeado a dos decimales. El resultado debería ser 4.76

Por último, calcula el número de palabras distintas del texto. Debe darte 90.

Prueba a calcular los tres números usando (a) listas y (b) diccionarios


Pregunta: ¿Qué diferencias existen entre las funciones replace() y remove()?

In [12]:
#RESOLUCIÓN CON LISTAS 
texto="""The Zen of Python, by Tim Peters   

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
"""

#aquí va tu código

In [13]:
#RESOLUCIÓN CON DICCIONARIOS

texto="""The Zen of Python, by Tim Peters   

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
"""

#aquí va tu código

## Ejercicio 5a.4

¿Os acordáis del porcentaje GC? Os he proporcionado una función que encapsula dicha funcionalidad.

Utilizando esa función y LISTAS, se pide que creéis un programa que procese las cadenas contenida en la variable fichFasta e imprima el porcentaje GC de cada secuencia de ADN. 



Un ejemplo de salida esperada es como sigue: 

<pre>
Porcentaje GC de secuencia  >HSBGPG Human gene for bone gla protein (BGP) : 63.52558895207149  ( 63.52558895207149 )
Porcentaje GC de secuencia  >HSGLTH1 Human theta 1-globin gene : 70.49019607843137  ( 70.49019607843137 )
</pre>



In [6]:
fichFasta="""
>HSBGPG Human gene for bone gla protein (BGP)
GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
ATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACC
ATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCC
CACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTG
GCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGG
GAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAAC
CAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACC
TGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTG
GCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTC
CACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGG
GTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTG
GCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGG
AGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCC
CCTGGAGCCCAGGAGGGAGGTGTGTGAGCTCAATCCGGACTGTGACGAGTTGGCTGACCACATCGGCTTTCAGGA
GGCCTATCGGCGCTTCTACGGCCCGGTCTAGGGTGTCGCTCTGCTGGCCTGGCCGGCAACCCCAGTTCTGCTCCT
CTCCAGGCACCCTTCTTTCCTCTTCCCCTTGCCCTTGCCCTGACCTCCCAGCCCTATGGATGTGGGGTCCCCATC
ATCCCAGCTGCTCCCAAATAAACTCCAGAAG
>HSGLTH1 Human theta 1-globin gene
CCACTGCACTCACCGCACCCGGCCAATTTTTGTGTTTTTAGTAGAGACTAAATACCATATAGTGAACACCTAAGA
CGGGGGGCCTTGGATCCAGGGCGATTCAGAGGGCCCCGGTCGGAGCTGTCGGAGATTGAGCGCGCGCGGTCCCGG
GATCTCCGACGAGGCCCTGGACCCCCGGGCGGCGAAGCTGCGGCGCGGCGCCCCCTGGAGGCCGCGGGACCCCTG
GCCGGTCCGCGCAGGCGCAGCGGGGTCGCAGGGCGCGGCGGGTTCCAGCGCGGGGATGGCGCTGTCCGCGGAGGA
CCGGGCGCTGGTGCGCGCCCTGTGGAAGAAGCTGGGCAGCAACGTCGGCGTCTACACGACAGAGGCCCTGGAAAG
GTGCGGCAGGCTGGGCGCCCCCGCCCCCAGGGGCCCTCCCTCCCCAAGCCCCCCGGACGCGCCTCACCCACGTTC
CTCTCGCAGGACCTTCCTGGCTTTCCCCGCCACGAAGACCTACTTCTCCCACCTGGACCTGAGCCCCGGCTCCTC
ACAAGTCAGAGCCCACGGCCAGAAGGTGGCGGACGCGCTGAGCCTCGCCGTGGAGCGCCTGGACGACCTACCCCA
CGCGCTGTCCGCGCTGAGCCACCTGCACGCGTGCCAGCTGCGAGTGGACCCGGCCAGCTTCCAGGTGAGCGGCTG
CCGTGCTGGGCCCCTGTCCCCGGGAGGGCCCCGGCGGGGTGGGTGCGGGGGGCGTGCGGGGCGGGTGCAGGCGAG
TGAGCCTTGAGCGCTCGCCGCAGCTCCTGGGCCACTGCCTGCTGGTAACCCTCGCCCGGCACTACCCCGGAGACT
TCAGCCCCGCGCTGCAGGCGTCGCTGGACAAGTTCCTGAGCCACGTTATCTCGGCGCTGGTTTCCGAGTACCGCT
GAACTGTGGGTGGGTGGCCGCGGGATCCCCAGGCGACCTTCCCCGTGTTTGAGTAAAGCCTCTCCCAGGAGCAGC
CTTCTTGCCGTGCTCTCTCGAGGTCAGGACGCGAGAGGAAGGCGC
>HSGLTH1 Human theta 1-globin2 gene
CCACTGCACTCACCGCACCCGGCCAATTTTTGTGTTTTTAGTAGAGACTAAATACCATATAGTGAACACCTAAGA
CGGGGGGCCTTGGATCCAGGGCGATTCAGAGGGCCCCGGTCGGAGCTGTCGGAGATTGAGCGCGCGCGGTCCCGG
GATCTCCGACGAGGCCCTGGACCCCCGGGCGGCGAAGCTGCGGCGCGGCGCCCCCTGGAGGCCGCGGGACCCCTG
GCCGGTCCGCGCAGGCGCAGCGGGGTCGCAGGGCGCGGCGGGTTCCAGCGCGGGGATGGCGCTGTCCGCGGAGGA
CCGGGCGCTGGTGCGCGCCCTGTGGAAGAAGCTGGGCAGCAACGTCGGCGTCTACACGACAGAGGCCCTGGAAAG
GTGCGGCAGGCTGGGCGCCCCCGCCCCCAGGGGCCCTCCCTCCCCAAGCCCCCCGGACGCGCCTCACCCACGTTC
CTCTCGCAGGACCTTCCTGGCTTTCCCCGCCACGAAGACCTACTTCTCCCACCTGGACCTGAGCCCCGGCTCCTC
ACAAGTCAGAGCCCACGGCCAGAAGGTGGCGGACGCGCTGAGCCTCGCCGTGGAGCGCCTGGACGACCTACCCCA
CGCGCTGTCCGCGCTGAGCCACCTGCACGCGTGCCAGCTGCGAGTGGACCCGGCCAGCTTCCAGGTGAGCGGCTG
CCGTGCTGGGCCCCTGTCCCCGGGAGGGCCCCGGCGGGGTGGGTGCGGGGGGCGTGCGGGGCGGGTGCAGGCGAG
TGAGCCTTGAGCGCTCGCCGCAGCTCCTGGGCCACTGCCTGCTGGTAACCCTCGCCCGGCACTACCCCGGAGACT
TCAGCCCCGCGCTGCAGGCGTCGCTGGACAAGTTCCTGAGCCACGTTATCTCGGCGCTGGTTTCCGAGTACCGCT
GAACTGTGGGTGGGTGGCCGCGGGATCCCCAGGCGACCTTCCCCGTGTTTGAGTAAAGCCTCTCCCAGGAGCAGC
CTTCTTGCCGTGCTCTCTCGAGGTCAGGACGCGAGAGGAAGGCGC
"""



def calculaGC(secuencia):
    #print(secuencia)
    totalCG=secuencia.count("G")+secuencia.count("C")
    totalACGT=secuencia.count("A")+secuencia.count("T")+secuencia.count("G")+secuencia.count("C")  
    return(totalCG*100/totalACGT)


def imprimeGCdeCadaSecuencia(secuenciaCompuesta):
    listaSec=secuenciaCompuesta.split(">")
    listaLimpia=[]
    for sec in listaSec:
        secLimpia=sec[sec.find("\n")+1:]
        secLimpia=secLimpia.replace("\n","")
            
        listaLimpia.append(secLimpia)
    while "" in listaLimpia:
        listaLimpia.remove("")
    for elemento in listaLimpia:
        print(calculaGC(elemento))
        
    
    
imprimeGCdeCadaSecuencia(fichFasta)


['', 'GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGTATAAACAGTGCTGGAGGCTGGCGGGGCAGGCCAGCTGAGTCCTGAGCAGCAGCCCAGCGCAGCCACCGAGACACCATGAGAGCCCTCACACTCCTCGCCCTATTGGCCCTGGCCGCACTTTGCATCGCTGGCCAGGCAGGTGAGTGCCCCCACCTCCCCTCAGGCCGCATTGCAGTGGGGGCTGAGAGGAGGAAGCACCATGGCCCACCTCTTCTCACCCCTTTGGCTGGCAGTCCCTTTGCAGTCTAACCACCTTGTTGCAGGCTCAATCCATTTGCCCCAGCTCTGCCCTTGCAGAGGGAGAGGAGGGAAGAGCAAGCTGCCCGAGACGCAGGGGAAGGAGGATGAGGGCCCTGGGGATGAGCTGGGGTGAACCAGGCTCCCTTTCCTTTGCAGGTGCGAAGCCCAGCGGTGCAGAGTCCAGCAAAGGTGCAGGTATGAGGATGGACCTGATGGGTTCCTGGACCCTCCCCTCTCACCCTGGTCCCTCAGTCTCATTCCCCCACTCCTGCCACCTCCTGTCTGGCCATCAGGAAGGCCAGCCTGCTCCCCACCTGATCCTCCCAAACCCAGAGCCACCTGATGCCTGCCCCTCTGCTCCACAGCCTTTGTGTCCAAGCAGGAGGGCAGCGAGGTAGTGAAGAGACCCAGGCGCTACCTGTATCAATGGCTGGGGTGAGAGAAAAGGCAGAGCTGGGCCAAGGCCCTGCCTCTCCGGGATGGTCTGTGGGGGAGCTGCAGCAGGGAGTGGCCTCTCTGGGTTGTGGTGGGGGTACAGGCAGCCTGCCCTGGTGGGCACCCTGGAGCCCCATGTGTAGGGAGAGGAGGGATGGGCATTTTGCACGGGGGCTGATGCCACCACGTCGGGTGTCTCAGAGCCCCAGTCCCCTACCCGGATCCCCTGGAGCCCAGGAGGGAG

## Ejercicio 5a.5

En genómica, es habitual encontrar regiones de ADN idénticas en genomas de organismos diferentes, lo que apunta a que se trata de una región conservada con muy probablemente una función muy parecida en ambos organismos.

En nuestro caso, entenderemos por motivo una secuencia *s* con longitud menor a la secuencia *t* donde querremos buscar ese motivo *len(s) < len(t)*. Por ejemplo, sean *t* y *s*:

t = GATATATGCATATACTT<br>
s = ATAT<br>

el motivo *s* aparece en las siguientes posiciones: 2, 4 y 10 (empezamos a contar en 1, aunque en Python se empiece a contar desde 0).

Escribid una función buscar_motivo(t,s) que devuelva en una lista las posiciones donde se ha encontrado el motivo en una determinada secuencia. Si no encuentra el motivo, devolverá una lista vacía. 

In [2]:
def buscar_motivo(t,s):
    #vuestro código
        

#código de ejemplo
t = 'AUGCUUCAGAAAGGUCUUACGUGCUUUA'
s = 'UGCU'
print(buscar_motivo(t, s)) #[2, 22]

t = 'AUGCUUCAGAAAGGUCUUACGU' 
s = 'U'
print(buscar_motivo(t, s))#[2, 5, 6, 15, 17, 18, 22]

t = 'AUGCUUCAGAAAGGUCUUACGU'
s = 'UGCG'
print(buscar_motivo(t, s)) #[]

t = 'GCGTAGTAGGTTCGCGCCTAGTTCCGCCGAAAAGACAAAGGAGAAGGGAATGCTCCTAGTAGTTTCAGTCTAGCAAACATGTTATAACGCTAACTGTGTGCTGCA'
s = 'AGA'
print(buscar_motivo(t, s)) #[33, 42]

t = "GATATATGCATATACTT"
s = "ATAT"
print(buscar_motivo(t, s)) #[2, 4, 10]

[2, 22]
[2, 5, 6, 15, 17, 18, 22]
[]
[33, 42]
[2, 10]


In [6]:
help(str.index)

Help on method_descriptor:

index(...)
    S.index(sub[, start[, end]]) -> int
    
    Like S.find() but raise ValueError when the substring is not found.



## Ejercicio 5a.6. 
Escribe una función llamada mapa que reciba una función y una lista, y devuelva la lista que resulta de aplicar la función recibida a cada uno de los elementos de la lista recibida. 


In [3]:
#aquí va vuestro código

#Pruebas: 
def multiplicaPorDos(a):
    return a*2

def sumarCinco(a):
    return a+5

lista=[1,2,3,4]
print(mapa(multiplicaPorDos,lista)) #[2, 4, 6, 8]
print(mapa(sumarCinco,lista)) #[6, 7, 8, 9]


NameError: name 'mapa' is not defined

## Ejercicio 5a.7. (Opt)
Generación de listas. Crea una lista con un número aleatorio de bases de ADN mediante una expresión generadora de listas. A continuación, únelas en una cadena con la función join().

Nota: recuerda, las bases se representan con el vocabulario ACGT

In [4]:
#aquí va vuestro código

# 5b.-DICCIONARIOS

## Ejercicio 5b.1
El formato PDB (Protein Data Bank, https://www.cgl.ucsf.edu/chimera/docs/UsersGuide/tutorials/pdbintro.html) se utiliza en química computacional para guardar información sobre moléculas en disco. Cada línea que empieza por ATOM representa un átomo de la molécula. Puedes leer más sobre este banco en https://es.wikipedia.org/wiki/Protein_Data_Bank. A continuación te mostramos los componentes de cada línea de tipo ATOM dentro de este fichero, que es lo que nos interesa en este ejercicio.


![Estructura Fichero PDB](estructuraFicheroPDB.png)

Veamos como ejemplo la primera parte de la especificación del Glucagon, una pequeña molécula de 29 amino ácidos en una sola cadena. El primer residuo es el aminoácido amino terminal, histidina, al que le sigue un residuo de serina y luego una glutamina. El tipo de cada átomo se encuentra en la última posición de cada línea. 

<pre>
ATOM      1  N   HIS A   1      49.668  24.248  10.436  1.00 25.00           N
ATOM      2  CA  HIS A   1      50.197  25.578  10.784  1.00 16.00           C
ATOM      3  C   HIS A   1      49.169  26.701  10.917  1.00 16.00           C
ATOM      4  O   HIS A   1      48.241  26.524  11.749  1.00 16.00           O
ATOM      5  CB  HIS A   1      51.312  26.048   9.843  1.00 16.00           C
ATOM      6  CG  HIS A   1      50.958  26.068   8.340  1.00 16.00           C
ATOM      7  ND1 HIS A   1      49.636  26.144   7.860  1.00 16.00           N
ATOM      8  CD2 HIS A   1      51.797  26.043   7.286  1.00 16.00           C
ATOM      9  CE1 HIS A   1      49.691  26.152   6.454  1.00 17.00           C
ATOM     10  NE2 HIS A   1      51.046  26.090   6.098  1.00 17.00           N
ATOM     11  N   SER A   2      49.788  27.850  10.784  1.00 16.00           N
ATOM     12  CA  SER A   2      49.138  29.147  10.620  1.00 15.00           C
ATOM     13  C   SER A   2      47.713  29.006  10.110  1.00 15.00           C
ATOM     14  O   SER A   2      46.740  29.251  10.864  1.00 15.00           O
ATOM     15  CB  SER A   2      49.875  29.930   9.569  1.00 16.00           C
ATOM     16  OG  SER A   2      49.145  31.057   9.176  1.00 19.00           O
ATOM     17  N   GLN A   3      47.620  28.367   8.973  1.00 15.00           N
ATOM     18  CA  GLN A   3      46.287  28.193   8.308  1.00 14.00           C
ATOM     19  C   GLN A   3      45.406  27.172   8.963  1.00 14.00           C
</pre>

Se pide que, dada la variable 'molecula', que contiene la cadena de texto que representa una molécula en formato PDB, defináis una función cuenta_atomos(molecula) que devuelva un diccionario con el número de átomos de cada tipo que tiene dicha molécula. 


In [45]:
molecula="""
ATOM      1  N   HIS A   1      49.668  24.248  10.436  1.00 25.00           N
ATOM      2  CA  HIS A   1      50.197  25.578  10.784  1.00 16.00           C
ATOM      3  C   HIS A   1      49.169  26.701  10.917  1.00 16.00           C
ATOM      4  O   HIS A   1      48.241  26.524  11.749  1.00 16.00           O
ATOM      5  CB  HIS A   1      51.312  26.048   9.843  1.00 16.00           C
ATOM      6  CG  HIS A   1      50.958  26.068   8.340  1.00 16.00           C
ATOM      7  ND1 HIS A   1      49.636  26.144   7.860  1.00 16.00           N
ATOM      8  CD2 HIS A   1      51.797  26.043   7.286  1.00 16.00           C
ATOM      9  CE1 HIS A   1      49.691  26.152   6.454  1.00 17.00           C
ATOM     10  NE2 HIS A   1      51.046  26.090   6.098  1.00 17.00           N
ATOM     11  N   SER A   2      49.788  27.850  10.784  1.00 16.00           N
ATOM     12  CA  SER A   2      49.138  29.147  10.620  1.00 15.00           C
ATOM     13  C   SER A   2      47.713  29.006  10.110  1.00 15.00           C
ATOM     14  O   SER A   2      46.740  29.251  10.864  1.00 15.00           O
ATOM     15  CB  SER A   2      49.875  29.930   9.569  1.00 16.00           C
ATOM     16  OG  SER A   2      49.145  31.057   9.176  1.00 19.00           O
ATOM     17  N   GLN A   3      47.620  28.367   8.973  1.00 15.00           N
ATOM     18  CA  GLN A   3      46.287  28.193   8.308  1.00 14.00           C
ATOM     19  C   GLN A   3      45.406  27.172   8.963  1.00 14.00           C
"""


molecula2="""
ATOM      8  CD2 HIS A   1      51.797  26.043   7.286  1.00 16.00           O
ATOM      9  CE1 HIS A   1      49.691  26.152   6.454  1.00 17.00           O
ATOM     10  NE2 HIS A   1      51.046  26.090   6.098  1.00 17.00           Ne
ATOM     11  N   SER A   2      49.788  27.850  10.784  1.00 16.00           Ne
ATOM     12  CA  SER A   2      49.138  29.147  10.620  1.00 15.00           He
ATOM     13  C   SER A   2      47.713  29.006  10.110  1.00 15.00           C
ATOM     14  O   SER A   2      46.740  29.251  10.864  1.00 15.00           O
ATOM     15  CB  SER A   2      49.875  29.930   9.569  1.00 16.00           C
ATOM     16  OG  SER A   2      49.145  31.057   9.176  1.00 19.00           O
ATOM     17  N   GLN A   3      47.620  28.367   8.973  1.00 15.00           Ne
ATOM     18  CA  GLN A   3      46.287  28.193   8.308  1.00 14.00           C
ATOM     19  C   GLN A   3      45.406  27.172   8.963  1.00 14.00           C
"""



250.1417


## Ejercicio 5b.2
Dada la misma molécula, crea ahora una función que almacene en un diccionario las coordenadas x,y,z (columnas 6, 7 y 8, empezando a contar desde 0) de cada átomo. La clave del diccionario será el número de átomo (columna 1, empezando a contar desde 0). A continuación, crea otra función que, dados los ids de dos átomos, calcule su distancia. 

Nota: la fórmula para calcular la distancia en un espacio tridimensional es: 
![DistanciaTridimensional](distanciaTridimensional.png)



In [16]:
def cargaDiccionario(m):
    d={}

    elementos=m.split("\n")
    while "" in elementos:
        elementos.remove("")
    for elemento in elementos:
        partes=elemento.split(" ")
        while "" in partes:
            partes.remove("")
        identif=int(partes[1])
        coordenadas=[]
        for i in range(6,9):
            coordenadas.append(float(partes[i]))
        d[identif]=coordenadas
    #aquí vuestro código
    return(d)


def calculaDistancia(m,idatomo1, idatomo2):
    dic=cargaDiccionario(m)
    x1,y1,z1=dic[idatomo1]
    x2,y2,z2=dic[idatomo2]
    d=((x2-x1)**2+(y2-y1)**2+(z2-z1)**2)**0.5
    return(d)
    #aquí vuestro código
    
    
    
molecula="""
ATOM      1  N   HIS A   1      49.668  24.248  10.436  1.00 25.00           N
ATOM      2  CA  HIS A   1      50.197  25.578  10.784  1.00 16.00           C
ATOM      3  C   HIS A   1      49.169  26.701  10.917  1.00 16.00           C
ATOM      4  O   HIS A   1      48.241  26.524  11.749  1.00 16.00           O
ATOM      5  CB  HIS A   1      51.312  26.048   9.843  1.00 16.00           C
ATOM      6  CG  HIS A   1      50.958  26.068   8.340  1.00 16.00           C
ATOM      7  ND1 HIS A   1      49.636  26.144   7.860  1.00 16.00           N
ATOM      8  CD2 HIS A   1      51.797  26.043   7.286  1.00 16.00           C
ATOM      9  CE1 HIS A   1      49.691  26.152   6.454  1.00 17.00           C
ATOM     10  NE2 HIS A   1      51.046  26.090   6.098  1.00 17.00           N
ATOM     11  N   SER A   2      49.788  27.850  10.784  1.00 16.00           N
ATOM     12  CA  SER A   2      49.138  29.147  10.620  1.00 15.00           C
ATOM     13  C   SER A   2      47.713  29.006  10.110  1.00 15.00           C
ATOM     14  O   SER A   2      46.740  29.251  10.864  1.00 15.00           O
ATOM     15  CB  SER A   2      49.875  29.930   9.569  1.00 16.00           C
ATOM     16  OG  SER A   2      49.145  31.057   9.176  1.00 19.00           O
ATOM     17  N   GLN A   3      47.620  28.367   8.973  1.00 15.00           N
ATOM     18  CA  GLN A   3      46.287  28.193   8.308  1.00 14.00           C
ATOM     19  C   GLN A   3      45.406  27.172   8.963  1.00 14.00           C
"""

#programa principal


dist=calculaDistancia(molecula,1,3)
print(dist)
dist=calculaDistancia(molecula,4,8)
print(dist)

{1: [49.668, 24.248, 10.436], 2: [50.197, 25.578, 10.784], 3: [49.169, 26.701, 10.917], 4: [48.241, 26.524, 11.749], 5: [51.312, 26.048, 9.843], 6: [50.958, 26.068, 8.34], 7: [49.636, 26.144, 7.86], 8: [51.797, 26.043, 7.286], 9: [49.691, 26.152, 6.454], 10: [51.046, 26.09, 6.098], 11: [49.788, 27.85, 10.784], 12: [49.138, 29.147, 10.62], 13: [47.713, 29.006, 10.11], 14: [46.74, 29.251, 10.864], 15: [49.875, 29.93, 9.569], 16: [49.145, 31.057, 9.176], 17: [47.62, 28.367, 8.973], 18: [46.287, 28.193, 8.308], 19: [45.406, 27.172, 8.963]}
2.549033346192238
5.72668019012761



## Ejercicio 5b.3
Dada una molécula representada por un string de átomos del estilo C9-H8-O4, crea una __función__ llamada calcula_masa_atomica(cadena) que reciba como parámetro dicha cadena y te devuelva su masa atómica. Por ejemplo: para una molécula C4-H3, la masa atómica sería 4*12.01+3.1.007825. 

Para ello, te proporcionamos un diccionario donde tienes como clave cada tipo de átomo, y como valor su masa atómica. 

In [6]:
dm={
'H':1.0079, 
'C':12.01,
'O':15.9994,
'N':14.0067,
'S':31.972071,
'P':30.9738,
'He':4.0026, 
'Li':6.941,
'Be':9.0122,
'Bo':10.811,
'F':18.9984,
'Ne':20.1797,
'Na':22.9897,
'Mg':24.305,
'Al':26.9815,
'Si':28.0855,
'Cl':35.453,
'K':39.0983,
'Ar':39.948,
'Ca':40.078,
'Sc':44.9559,
'Ti':47.867,
'V':50.9415,
'Cr':51.9961,
'Mn':54.938,
'Fe':55.845,
'Ni':58.6934,
'Co':58.9332,
'Cu':63.546,
'Zn':65.39,
'Ga':69.723,
'Ge':72.64,
'As':74.9216,
'Se':78.96,
'Br':79.904,
'Kr':83.8,
'Rb':85.4678,
'Sr':87.62,
'Y':88.9059,
'Zr':91.224,
'Nb':92.9064,
'Mo':95.94,
'Tc':98,
'Ru':101.07,
'Rh':102.9055,
'Pd':106.42,
'Ag':107.8682,
'Cd':112.411,
'In':114.818,
'Sn':118.71,
'Sb':121.76,
'I':126.9045,
'Te':127.6,
'Xe':131.293,
'Cs':132.9055,
'Ba':137.327,
'La':138.9055,
'Ce':140.116,
'Pr':140.9077,
'Nd':144.24,
'Pm':145,
'Sm':150.36,
'Eu':151.964,
'Gd':157.25,
'Tb':158.9253,
'Dy':162.5,
'Ho':164.9303,
'Er':167.259,
'Tu':168.9342,
'Yb':173.04,
'Lu':174.967,
'Hf':178.49,
'Ta':180.9479,
'W':183.84,
'Re':186.207,
'Os':190.23,
'Ir':192.217,
'Pt':195.078,
'Au':196.9665,
'Hg':200.59,
'Ta':204.3833,
'Pb':207.2,
'Bi':208.9804,
'Po':209,
'At':210,
'Rd':222,
'Fr':223,
'Ra':226,
'Ac':227,
'Pa':231.0359,
'Th':232.0381,
'Np':237,
'U':238.0289,
'Am':243,
'Pu':244,
'Cu':247,
'Bk':247,
'Cf':251,
'Es':252,
'Fm':257,
'Md':258,
'No':259,
'Rf':261,
'Lr':262,
'Db':262,
'Bh':264,
'Sg':266,
'Mt':268,
'Hs':277}


#Aquí va vuestra función

def calcula_masa_atomica(molecula):
    try:
        elementos=[]
        elementos=molecula.split("-")
        masa=0
        ltipos=[]
        for elemento in elementos:
            tipo=""
            numveces=""
            for c in elemento:
                if c in "0123456789":# if c.isdigit()
                    numveces=numveces+c
                else:
                    tipo=tipo+c
                    
            if tipo in ltipos:
                raise Exception("Elemento repetido")
            else:
                ltipos.append(tipo)
            if numveces=="":
                numveces="1"
            masa=masa+dm[tipo]*int(numveces)
        return(masa)
    except KeyError:
        print("Error de clave: elemento no válido",elemento)
        return(0)
    except Exception as ex:
        print("Error genérico: ",format(ex))
        return(0)
    
print("La masa atómica de H3-H2 es ",calcula_masa_atomica("H3-H2")) #error mismo elemento aparece más de una vez    
print("La masa atómica de H2O es ",calcula_masa_atomica("H2O")) #error falta separador  
print("La masa atómica de Tz1-H3 ",calcula_masa_atomica("Tz1-H3")) #error elemento no existe   
print("La masa atómica de H3 es ",calcula_masa_atomica("H3")) #sólo hay un elemento  
print("La masa atómica de N_H3 es ",calcula_masa_atomica("N_H3")) #separador erróneo    
print("La masa atómica del amoníaco (N-H3) es ",calcula_masa_atomica("N-H3")) #amoníaco: 17.0304
print("La masa atómica de la celulosa (C6-H10-O5) es ",calcula_masa_atomica("C6-H10-O5")) #celulosa: 162.13600000000002
print("La masa atómica de la glucosa (C6-H12-O6) es ",calcula_masa_atomica("C6-H12-O6")) #glucosa: 180.1512
print("La masa atómica del cloruro de sodio (Na-Cl) es ",calcula_masa_atomica("Na-Cl")) #cloruro de sodio: 58.4427

Error genérico:  Elemento repetido
La masa atómica de H3-H2 es  0
Error de clave: elemento no válido H2O
La masa atómica de H2O es  0
Error de clave: elemento no válido Tz1
La masa atómica de Tz1-H3  0
La masa atómica de H3 es  3.0237
Error de clave: elemento no válido N_H3
La masa atómica de N_H3 es  0
La masa atómica del amoníaco (N-H3) es  17.0304
La masa atómica de la celulosa (C6-H10-O5) es  162.13600000000002
La masa atómica de la glucosa (C6-H12-O6) es  180.1512
La masa atómica del cloruro de sodio (Na-Cl) es  58.4427


## Ejercicio 5b.4

¿Os acordáis de la estructura primaria, secundaria y terciaria del ADN?


<img src="https://upload.wikimedia.org/wikipedia/commons/a/ae/DNA_chemical_structure_es.svg" alt="ADN" style="width: 300px;"/>

En este ejercicio os volvemos a pedir que, dada una secuencia con bases A, C, G o T, imprimáis por pantalla su secuencia complementaria. Esta secuencia se calcula como los elementos complementarios de cada elemento de la secuencia original ordenados en sentido inverso. 

Ejemplo: 
<pre>
Secuencia:      GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
Complementaria: ACCCTCTGGGCTGTGCCCAGCGATGAGGAGGGGCATGCCTGACCATGGTGCGGGCGGGTCTAGGGGGAATCTGCC 
</pre>

A continuación os mostramos el código necesario para realizar esta operación con operaciones simples sobre cadenas. Tenéis que implementar una función que haga lo mismo USANDO UN DICCIONARIO. 

¿Cuál de las dos implementaciones es más eficiente? (Pista: podéis usar el magic command %timeit para averiguarlo)

In [5]:


def seqComplementariaConCadenas(cadena):
    reverso_complementario=""
    for c in cadena:
        if (c=="T"):
            cc="A"
        elif(c=="A"):
            cc="T"
        elif (c=="C"):
            cc="G"
        elif (c=="G"):
            cc="C"
        reverso_complementario=cc+reverso_complementario #al unir al final estoy haciendo el reverso
    return(reverso_complementario)

def seqComplementariaConDiccionario(cadena):
    pass
    #aquí va vuestro código


cadena="GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT"
print(cadena)

print(seqComplementariaConCadenas(cadena))
%timeit seqComplementariaConCadenas(cadena)
print(seqComplementariaConDiccionario(cadena))
%timeit seqComplementariaConDiccionario(cadena)





GGCAGATTCCCCCTAGACCCGCCCGCACCATGGTCAGGCATGCCCCTCCTCATCGCTGGGCACAGCCCAGAGGGT
ACCCTCTGGGCTGTGCCCAGCGATGAGGAGGGGCATGCCTGACCATGGTGCGGGCGGGTCTAGGGGGAATCTGCC
10000 loops, best of 3: 18.5 µs per loop
None
10000000 loops, best of 3: 109 ns per loop


## Ejercicio 5b.5

En la transcripción de ADN a proteínas el concepto de codón juega un papel importante. Un codón es un triplete de nucleótidos, y cada triplete codifica un aminoácido (residuo), aunque un aminoácido puede estar codificado por varios codones. Existe, tres codones especiales, de parada o terminación, que no determinan ningún aminoácido, pero que indica que la transcripción ha de terminar y la proteína está completa (codones "*"). 

Tenéis disponible una tabla de codones en https://en.wikipedia.org/wiki/DNA_codon_table. Esta tabla se ha codificado como un diccionario Python. 

Dada una secuencia de ADN, crea una función que reciba una cadena que representa la secuencia de ADN y que devuelva la cadena que representa la proteína transcrita. Queremos que la función pare de traducir cuando se encuentre un codón de stop, y devuelva None si no reconoce cualquiera de los codones que componen la secuencia de ADN. 


In [90]:
diccionarioCodones={"TTT":"F","TTC":"F","TTA":"L","TTG":"L", "CTT":"L", "CTC":"L", "CTA":"L", "CTG":"L","ATT":"I","ATC":"I", 
                    "ATA":"I", "ATG":"M", "GTT":"V", "GTC":"V", "GTA":"V", "GTG":"V", "TCT":"S","TCC":"S", "TCA":"S", "TCG":"S", 
                    "CCT":"P","CCC":"P", "CCA":"P", "CCG":"P","ACT":"T", "ACC":"T", "ACA":"T", "ACG":"T","GCT":"A", "GCC":"A",
                    "GCA":"A", "GCG":"A", "TAT":"Y","TAC":"Y","TAA":"*","TAG":"*","CAT":"H", "CAC":"H","CAA":"Q","CAG":"Q",
                    "AAT":"N","AAC":"N","AAA":"L","AAG":"L","GAT":"D","GAC":"D","GAA":"E","GAG":"E","TGT":"C","TGC":"C",
                    "TGA":"*","TGG":"W","CGT":"R","CGC":"R", "CGA":"R", "CGG":"R","AGT":"S", "AGC":"S",
                    "AGA":"R", "AGG":"R","GGT":"G","GGC":"G", "GGA":"G", "GGG":"G"}




#aquí va vuestro código



#ejemplo
seqADN1="AGTGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGCTGGGATTT"
print(seqADN1)
print("Salida:",ADNtoProtein(seqADN1),"\n") #SDHIIIDRVNLEDLLGF
seqADN2="AGTGATTGAATAATAATTGATCGAGTTAATCTGGAGGATCTGCTGGGATTT"
print(seqADN2)
print("Salida:",ADNtoProtein(seqADN2),"\n") #SD
print("Salida:",ADNtoProtein("prueba"),"\n") #None



AGTGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGCTGGGATTT
Salida: SDHIIIDRVNLEDLLGF 

AGTGATTGAATAATAATTGATCGAGTTAATCTGGAGGATCTGCTGGGATTT
Salida: SD 

No reconozco la cadena como una cadena de ADN válida
Salida: None 



## Ejercicio 5b.6

Siguiendo con el mismo tema de los codones, nos han pasado la siguiente lista, que contiene todos los codones con sus correspondientes aminoácidos:
<pre>
['AGT:S', 'ACG:T', 'GCT:A', 'TCT:S', 'TCA:S', 'CGA:R', 'AAT:N', 'ACA:T', 'CTC:L', 'GCG:A', 'ACT:T', 'ATA:I', 'GGC:G', 'TTC:F', 'TAA:*', 'TGA:*', 'CCC:P', 'GTA:V', 'TAT:Y', 'ATT:I', 'GGT:G', 'ATG:M', 'CAT:H', 'CTT:L', 'CAG:Q', 'TGG:W', 'CTG:L', 'AAG:L', 'CTA:L', 'ATC:I', 'AGA:R', 'CAC:H', 'AGG:R', 'GTG:V', 'TTT:F', 'AGC:S', 'GCC:A', 'GTT:V', 'GGA:G', 'GCA:A', 'GAT:D', 'AAA:L', 'CCG:P', 'TGC:C', 'GAG:E', 'TTG:L', 'GAC:D', 'TCC:S', 'CGC:R', 'GGG:G', 'CAA:Q', 'CGT:R', 'AAC:N', 'CCT:P', 'ACC:T', 'TAG:*', 'GTC:V', 'GAA:E', 'TTA:L', 'CGG:R', 'TGT:C', 'TAC:Y', 'TCG:S', 'CCA:P']
</pre>
Nos piden codificar una función llamada codonesPorAminoacido() que reciba como parámetro la lista de codones y nos devuelva un diccionario donde la clave es el aminoácido y el valor es una tupla (veces,porcentaje) donde veces es el número de codones que codifican ese aminoácido y porcentaje es el porcentaje de codones frente al total.

Además, deseamos imprimir el diccionario resultante ordenado: 
* Alfabéticamente
* Por número de codones que codifican ese aminoácido (veces), de mayor a menor


Salida esperada: 

Diccionario Aminoácidos con tuplas ordenadas alfabéticamente por nombre de aminoácido
 [('*', (3, 4.6875)), ('A', (4, 6.25)), ('C', (2, 3.125)), ('D', (2, 3.125)), ('E', (2, 3.125)), ('F', (2, 3.125)), ('G', (4, 6.25)), ('H', (2, 3.125)), ('I', (3, 4.6875)), ('L', (8, 12.5)), ('M', (1, 1.5625)), ('N', (2, 3.125)), ('P', (4, 6.25)), ('Q', (2, 3.125)), ('R', (6, 9.375)), ('S', (6, 9.375)), ('T', (4, 6.25)), ('V', (4, 6.25)), ('W', (1, 1.5625)), ('Y', (2, 3.125))]<br>
Diccionario Aminoácidos con tuplas ordenadas por número de aminoácidos de mayor a menor
 [('L', (8, 12.5)), ('R', (6, 9.375)), ('S', (6, 9.375)), ('T', (4, 6.25)), ('G', (4, 6.25)), ('A', (4, 6.25)), ('V', (4, 6.25)), ('P', (4, 6.25)), ('I', (3, 4.6875)), ('*', (3, 4.6875)), ('C', (2, 3.125)), ('D', (2, 3.125)), ('F', (2, 3.125)), ('E', (2, 3.125)), ('Q', (2, 3.125)), ('N', (2, 3.125)), ('Y', (2, 3.125)), ('H', (2, 3.125)), ('W', (1, 1.5625)), ('M', (1, 1.5625))]



In [11]:
listaCodones=['AGT:S', 'ACG:T', 'GCT:A', 'TCT:S', 'TCA:S', 'CGA:R', 'AAT:N', 'ACA:T', 'CTC:L', 'GCG:A', 'ACT:T', 
              'ATA:I', 'GGC:G', 'TTC:F', 'TAA:*', 'TGA:*', 'CCC:P', 'GTA:V', 'TAT:Y', 'ATT:I', 'GGT:G', 'ATG:M', 
              'CAT:H', 'CTT:L', 'CAG:Q', 'TGG:W', 'CTG:L', 'AAG:L', 'CTA:L', 'ATC:I', 'AGA:R', 'CAC:H', 'AGG:R', 
              'GTG:V', 'TTT:F', 'AGC:S', 'GCC:A', 'GTT:V', 'GGA:G', 'GCA:A', 'GAT:D', 'AAA:L', 'CCG:P', 'TGC:C', 
              'GAG:E', 'TTG:L', 'GAC:D', 'TCC:S', 'CGC:R', 'GGG:G', 'CAA:Q', 'CGT:R', 'AAC:N', 'CCT:P', 'ACC:T', 
              'TAG:*', 'GTC:V', 'GAA:E', 'TTA:L', 'CGG:R', 'TGT:C', 'TAC:Y', 'TCG:S', 'CCA:P']


#en dos pasos
d={}
totalCodones=len(listaCodones)
for elemento in listaCodones:
    codon,amino=elemento.split(":")
    d[amino]=d.get(amino,0)+1

for clave in d.keys():
    d[clave]=(d[clave],d[clave]/totalCodones)

print(d)



#trabajando directamente con tuplas dentro del diccionario
d={}
totalCodones=len(listaCodones)
for elemento in listaCodones:
    codon,amino=elemento.split(":")
    num,porc=d.get(amino,(0,0))
    num=num+1
    porc=num/totalCodones
    d[amino]=(num,porc)

print(d)


#aquí va vuestro código

{'I': (3, 0.046875), 'G': (4, 0.0625), 'W': (1, 0.015625), 'C': (2, 0.03125), 'T': (4, 0.0625), 'N': (2, 0.03125), 'Q': (2, 0.03125), 'P': (4, 0.0625), 'H': (2, 0.03125), 'E': (2, 0.03125), 'L': (8, 0.125), 'A': (4, 0.0625), 'S': (6, 0.09375), 'Y': (2, 0.03125), 'F': (2, 0.03125), 'M': (1, 0.015625), 'R': (6, 0.09375), '*': (3, 0.046875), 'V': (4, 0.0625), 'D': (2, 0.03125)}
{'I': (3, 0.046875), 'G': (4, 0.0625), 'W': (1, 0.015625), 'C': (2, 0.03125), 'T': (4, 0.0625), 'N': (2, 0.03125), 'Q': (2, 0.03125), 'P': (4, 0.0625), 'H': (2, 0.03125), 'E': (2, 0.03125), 'L': (8, 0.125), 'A': (4, 0.0625), 'S': (6, 0.09375), 'Y': (2, 0.03125), 'F': (2, 0.03125), 'M': (1, 0.015625), 'R': (6, 0.09375), '*': (3, 0.046875), 'V': (4, 0.0625), 'D': (2, 0.03125)}


In [18]:
tupla=(1,2,4,1,5,6,7)
print(tupla)

lista=[]
lista.append(tupla)
print(lista)
tupla2=(3,4)
lista.append(tupla2)
print(lista)

(1, 2, 4, 1, 5, 6, 7)
[(1, 2, 4, 1, 5, 6, 7)]
[(1, 2, 4, 1, 5, 6, 7), (3, 4)]


## Ejercicio 5b.7

Escribe un programa que vaya solicitando al usuario que ingrese nombres de personas.
Si el nombre se encuentra en la agenda (implementada con un diccionario), el programa debe mostrar el teléfono y preguntar si se desea modificar el teléfono, introducir otro nombre para buscar su teléfono o salir del programa. 
Si el nombre no se encuentra, el programa debe permitir ingresar el teléfono y guardarlo en el diccionario (la agenda). 



In [ ]:
agenda={"Rosa Berauto":"964562343","José Pérez":"965234812","Pilar Rodríguez":"83726191"}
#aquí va tu código


## Ejercicio 5b.8
Retomemos el formato PDB del ejercicio 5b.3. Implementad una función cuentaAminos(molecula) que calcule el número de aminoácidos que contiene la molécula que se le pase como parámetro. 

In [36]:
molecula="""
ATOM      1  N   HIS A   1      49.668  24.248  10.436  1.00 25.00           N
ATOM      2  CA  HIS A   1      50.197  25.578  10.784  1.00 16.00           C
ATOM      3  C   HIS A   1      49.169  26.701  10.917  1.00 16.00           C
ATOM      4  O   HIS A   1      48.241  26.524  11.749  1.00 16.00           O
ATOM      5  CB  HIS A   1      51.312  26.048   9.843  1.00 16.00           C
ATOM      6  CG  HIS A   1      50.958  26.068   8.340  1.00 16.00           C
ATOM      7  ND1 HIS A   1      49.636  26.144   7.860  1.00 16.00           N
ATOM      8  CD2 HIS A   1      51.797  26.043   7.286  1.00 16.00           C
ATOM      9  CE1 HIS A   1      49.691  26.152   6.454  1.00 17.00           C
ATOM     10  NE2 HIS A   1      51.046  26.090   6.098  1.00 17.00           N
ATOM     11  N   SER A   2      49.788  27.850  10.784  1.00 16.00           N
ATOM     12  CA  SER A   2      49.138  29.147  10.620  1.00 15.00           C
ATOM     13  C   SER A   2      47.713  29.006  10.110  1.00 15.00           C
ATOM     14  O   SER A   2      46.740  29.251  10.864  1.00 15.00           O
ATOM     15  CB  SER A   2      49.875  29.930   9.569  1.00 16.00           C
ATOM     16  OG  SER A   2      49.145  31.057   9.176  1.00 19.00           O
ATOM     17  N   GLN A   3      47.620  28.367   8.973  1.00 15.00           N
ATOM     18  CA  GLN A   3      46.287  28.193   8.308  1.00 14.00           C
ATOM     19  C   GLN A   3      45.406  27.172   8.963  1.00 14.00           C
"""

# Aquí va vuestro código

#código de prueba
print(cuentaAminos(molecula))

NameError: name 'cuentaAminos' is not defined

# 5c.-TUPLAS

## Ejercicio 5c.1. 
EcoRI es una enzima de restricción muy empleada en todos los laboratorios del mundo para cortar una hebra de ADN en dos partes. EcoRI reconoce el motivo GAATTC, y separa la secuencia inicial de ADN por la posición tras la primera G (G-AATTC) Dada una cadena que representa una hebra de ADN, escribid una función llamada cortaHebraConEcoRI() que reciba como parámetro la hebra de ADN y devuelva una tupla que contenga las dos partes de la hebra de ADN cortada por el motivo GAATTC. Si no lo encuentra, la función debe devolver la hebra completa como primer elemento de la tupla, y None como segundo. 

Salida esperada con la hebra proporcionada como ejemplo: 
('ACTGATCGATTACGTATAGTAG', 'AATTCTATCATACATATATATCGATGCGTTCATTAC')

In [41]:
#aquí vuestro código


hebra = 'ACTGATCGATTACGTATAGTAGAATTCTATCATACATATATATCGATGCGTTCATTAC'
print(cortaHebraConEcoRI(hebra))

('ACTGATCGATTACGTATAGTAG', 'AATTCTATCATACATATATATCGATGCGTTCATTAC')


## Ejercicio 5c.2.

Escribe una función que indique si dos fichas de dominó ‘encajan’ o no. La función recibirá dos parámetros, que se corresponderán a dos tuplas, una por cada ficha de dominó. A continuación, escribe un programa principal que genere las fichas del dominó,  de entre ellas seleccione 10 veces dos al azar e imprima si las dos fichas seleccionadas encajan o no. 




In [ ]:
#aquí va vuestro código

print(encajan((1,1)(0,0))) #False
print(encajan((1,4)(0,1))) #True
print(encajan((1,6)(4,6))) #True
print(encajan((1,6)(3,4))) #False
print(encajan((1,1)(5,6))) #False
print(encajan((3,5)(0,3))) #True
print(encajan((3,6)(2,3))) #True
print(encajan((0,0)(0,1))) #True
print(encajan((0,1)(4,5))) #False
print(encajan((2,2)(5,6))) #False

## Ejercicio 5c.3. 

A partir del código del ejercicio anterior, escribe un programa que juegue automáticamente al dominó. Para ello: 
*	Genera la lista de 28 fichas (una lista con 28 tuplas)
*	Asume cuatro jugadores
*	Asigna 7 fichas al azar a cada jugador
*	Comienza con el jugador que tenga el (6, 6)
*	Crea una lista que represente el juego encima de la mesa. Cuando coloques las fichas (las insertes en esa lista, al principio o al final), cuida de colocarlas en el orden que toca, para saber qué número hay en cada extremo.  Las dobles da igual en qué orden las enlaces. 
*	El juego terminará cuando alguno de los jugadores se quede sin fichas o no se puedan colocar más fichas. En este caso gana el jugador que sume menos puntos. 


Pistas: 
*	lista.append(elemento) añade un elemento al final de la lista.
*	lista.insert(pos,elemento) añade un nuevo valor en la posición k (recuerda que la primera posición siempre es la 0) y desplaza un lugar el resto de la lista. 
*	Lista.remove(elemento) borra el elemento de la lista. El valor debe existir.
*	Para ver si un elemento se encuentra en una lista puedes utilizar el operador in


In [ ]:
#aquí vuestro código

## Ejercicio 5c.4.

Escribe una función dime_fecha() que recupere la hora actual del sistema (time.localtime()) y transforme la tupla de tiempo devuelta por dicha función en una cadena en formato “Hoy es  diaSem, dd de Mes de yyyy. Son las hh:mm.”  que imprimirá por pantalla. 

E.g. de salida: Hoy es Lunes, 29 de Mar de 2018. Son las 14:02.
 

Nota: La salida de time.localtime() es un objeto time.struct_time(tm_year=2016, tm_mon=2, tm_mday=29, tm_hour=13, tm_min=49, tm_sec=35, tm_wday=0, tm_yday=60, tm_isdst=0)

Struct_time es una named_tuple. Podéis ver cómo se trabaja con ellas en https://docs.python.org/3.6/glossary.html#term-named-tuple



In [ ]:
#aquí va tu código

## Ejercicio 5c.5.

Escribe una función diaSiguiente que pida al usuario un día, un mes y un año –donde día y año son enteros, y mes es una cadena (Ene,Feb,… Dic). A partir de esa entrada, la función debe calcular el día siguiente y y devolverlo en una tupla con formato(dia-entero,mes-cadena,año-entero). 

Pistas: es posible crear una fecha a partir de un valor de día, mes y año con la función date(año, mes, día) . Esta función se encuentra en el módulo datetime. También es posible sumar 1 a una fecha usando la función timedelta(days=1) como sigue: diaSig=fecha+timedelta(days=1)

Esta función también se encuentra en el módulo datetime. 

Ejemplos de ejecución: 
![Ejemplos de salida](5c-4-salida.png)

In [ ]:
#aquí va tu código

## Ejercicio 5c.6.

Escribe una función empaquetar que, dada una lista que le pasamos como parámetro, nos devuelva otra lista donde la repetición de valores se indica mediante tuplas. 

Ejemplo: 
<pre>
lista=[1,2,2,2,3,3,3,3,3,3,3]
print(empaquetar(lista)) #[(1,1),(2,3),(3,6)]
</pre>

In [ ]:
#aquí va vuestro código


#ejemplos de ejecución
lista1=[1,1,1,1,1,1,2,2,2,2,2,3,3,3,3,4,5,5,5,5,6,6,6,7,7,8,8,8,8,8,8,9]
lista2=[1,2,3,3,3,3,3,3,3,3,3,3,3,4,5,6,7,7,7,7,7,7,7,7,7,7,7,8]
print(empaquetar(lista1))
print(empaquetar(lista2))

## Ejercicio 5c.7.

Modifica la función empaquetar para que acepte un parámetro que indique si tenemos que tener en cuenta que las ocurrencias repetidas sean consecutivas para contarlas dentro de una misma tupla (consec=True) o no. Por defecto consec=False

Ejemplo: 
<pre>
lista=[1,2,2,2,3,3,3,3,3,3,3,1,1,1,1]
print(empaquetar(lista)) #[(1,5),(2,3),(3,6)]
print(empaquetar(lista,False)) #[(1,5),(2,3),(3,6)]
print(empaquetar(lista,True)) #[(1,1),(2,3),(3,6),(1,4)]
</pre>

In [2]:
def empaquetar(listanum,consec=False):
    lista=[]
    if consec==False:
        dcont={}
        for elemento in listanum:
            dcont[elemento]=dcont.get(elemento,0)+1
        for clave in dcont:
            lista.append((clave,dcont[clave]))

    else: 
        actual=None
        contador=0
        for elemento in listanum:
            print(elemento)
            if actual==None:
                contador=1
                actual=elemento
            elif elemento!=actual: 
                lista.append((actual,contador))
                contador=1
                actual=elemento
            else: #elemento repetido
                contador=contador+1
        lista.append((actual,contador))

    return(lista)
    
lista=[1,2,2,2,3,3,3,3,3,3,3,1,1,1,1]
print(empaquetar(lista,True))
    

1
2
2
2
3
3
3
3
3
3
3
1
1
1
1
[(1, 1), (2, 3), (3, 7), (1, 4)]


# Ejercicio 5c.8.

Crea un programa que use una named tuple para calcular la distancia euclídea entre dos puntos

Pista: namedtuple se encuentra definida en el módulo collections:
from collections import namedtuple


Pista: podéis ver explicado este ejemplo en http://librosweb.es/libro/algoritmos_python/capitulo_7/


In [ ]:
#aquí vuestro código

## Ejercicio 5c.9.

Nos han pasado el siguiente programa principal de una máquina expendedora de medicamentos:

<pre>
listaMedicamentos=[("Ibuprofeno",1.97),("Paracetamol",2.50),("Strepsils",9.20),("Agua Oxigenada",3.83),("Betadine",6.71)]

def principal(): 
    numlist=dameOpcionMenu(listaMedicamentos)
    if (numlist<=len(listaMedicamentos)): #ha elegido algún medicamento
        print("Has elegido comprar ",listaMedicamentos[numlist-1][0])
        print("El precio de",listaMedicamentos[numlist-1][0], "es",listaMedicamentos[numlist-1][1])
        listaDinero=generaDineroIntroducido(listaMedicamentos[numlist-1][1])
        total=calculaTotalDineroIntroducido(listaDinero) 
        print("Has introducido ",total," euros")

        cambio=calculaCambio(total,listaMedicamentos[numlist-1][1])
        print("Tu cambio es de ",cambio, "euros")
    else:
        print("Gracias por confiar en nosotros. Esperamos volver a verte pronto.")


        
principal()
</pre>


Define las funciones necesarias para que el programa funcione correctamente. 
Ejemplo de ejecución: 

<pre>
Medicamentos
1 .- Ibuprofeno --> 1.97
2 .- Paracetamol --> 2.5
3 .- Strepsils --> 9.2
4 .- Agua Oxigenada --> 3.83
5 .- Betadine --> 6.71
6 .- Calcelar operación.
Por favor, elige la opción que deseas comprar...3
Has elegido comprar  Strepsils
El precio de Strepsils es 9.2
Dinero introducido:  [0.1, 10]
Has introducido  10.1  euros
Tu cambio es de  0.9 euros
</pre>

In [1]:
#Aquí va tu código